File: android/app/src/main/java/com/laneshadow/ui/components/molecules/DiscoveryFilterBar.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme
import com.laneshadow.ui.components.atoms.Chip

/**
 * Draw a bottom border on the modifier
 */
private fun Modifier.drawBottomBorder(color: Color, thickness: Dp) = this.drawBehind {
    val strokeWidth = thickness.toPx()
    val y = size.height - strokeWidth / 2
    drawLine(
        color = color,
        start = Offset(0f, y),
        end = Offset(size.width, y),
        strokeWidth = strokeWidth,
    )
}

/**
 * Route archetype enum
 *
 * Represents different types of motorcycle routes.
 * Following RN contract from react-native/components/ui/discovery-filter-bar.tsx
 */
enum class RouteArchetype {
    All,
    Twisties,
    Scenic,
    Technical,
    Cruising,
    Sport,
    Adventure,
}

/**
 * DiscoveryFilterBar molecule component
 *
 * Horizontal scrollable chip bar with glassmorphic design for filtering routes by archetype.
 * Following RN wrapper API from react-native/components/ui/discovery-filter-bar.tsx
 *
 * Design tokens (from spec):
 * - Surface background: theme.colors.surface.default with 0.8f alpha (80% opacity)
 * - Border bottom: theme.colors.border.default with 0.13f alpha (20% of 0.65 base opacity)
 * - Padding: theme.space.md top/bottom, theme.space.lg horizontal
 * - Spacing: theme.space.sm between chips
 * - Uses existing Chip component from atoms package
 *
 * Behavior:
 * - "All" chip clears selection (when selected, no other archetypes are selected)
 * - Other archetypes toggle (multiple can be selected)
 * - Count displayed next to each label: "Scenic (12)"
 *
 * @param selectedArchetypes List of currently selected route archetypes
 * @param onArchetypeChange Callback when archetype selection changes
 * @param counts Map of archetype to count (for display labels)
 * @param testID Optional test ID for UI testing
 * @param modifier Modifier for the container
 */
@Composable
fun DiscoveryFilterBar(
    selectedArchetypes: List<RouteArchetype>,
    onArchetypeChange: (List<RouteArchetype>) -> Unit,
    counts: Map<RouteArchetype, Int>,
    testID: String? = null,
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    // Glassmorphic surface background: 80% opacity
    val surfaceColor = theme.colors.surface.default.copy(alpha = 0.8f)

    // Border color: 20% opacity of base border (0.13f)
    val borderColor = theme.colors.border.default.copy(alpha = 0.13f)

    // Container with glassmorphic styling
    val containerModifier = modifier
        .background(color = surfaceColor)
        .drawBottomBorder(borderColor, 1.dp)
        .padding(
            horizontal = theme.space.lg,
            vertical = theme.space.md,
        )

    // Horizontal scrolling row of chips
    LazyRow(
        modifier = containerModifier,
        horizontalArrangement = Arrangement.spacedBy(theme.space.sm),
    ) {
        // Get all archetype values in order
        val allArchetypes = RouteArchetype.entries

        // Add chip for each archetype
        for (archetype in allArchetypes) {
            item(key = archetype.name) {
                val isSelected = selectedArchetypes.contains(archetype)
                val count = counts[archetype] ?: 0
                val label = when (archetype) {
                    RouteArchetype.All -> "All"
                    else -> "${archetype.name} ($count)"
                }

                Chip(
                    label = label,
                    selected = isSelected,
                    onPress = {
                        handleArchetypePress(
                            archetype = archetype,
                            currentSelection = selectedArchetypes,
                            onArchetypeChange = onArchetypeChange,
                        )
                    },
                    testID = testID?.let { "$it-${archetype.name}" },
                )
            }
        }
    }
}

/**
 * Handle archetype chip press
 *
 * - "All" clears selection (selects only All)
 * - Selecting another archetype deselects "All"
 * - Toggling an archetype adds/removes it from selection
 * - If all specific archetypes are deselected, "All" becomes selected
 */
private fun handleArchetypePress(
    archetype: RouteArchetype,
    currentSelection: List<RouteArchetype>,
    onArchetypeChange: (List<RouteArchetype>) -> Unit,
) {
    when (archetype) {
        RouteArchetype.All -> {
            // "All" clears selection - only select All
            onArchetypeChange(listOf(RouteArchetype.All))
        }
        else -> {
            // Toggle the archetype
            val newSelection = if (currentSelection.contains(archetype)) {
                // Remove the archetype
                currentSelection - archetype
            } else {
                // Add the archetype and remove "All" if present
                (currentSelection - RouteArchetype.All) + archetype
            }

            // If no specific archetypes selected, default to "All"
            if (newSelection.isEmpty()) {
                onArchetypeChange(listOf(RouteArchetype.All))
            } else {
                onArchetypeChange(newSelection)
            }
        }
    }
}

File: ios/LaneShadow/Views/Molecules/DiscoveryFilterBar.swift
import LaneShadowTheme
import SwiftUI

// MARK: - Route Archetype Enum

/**
 * Route archetype enum
 *
 * Following RN API from react-native/components/discovery/discovery-filter-bar.tsx
 * Represents different types of motorcycle routes for filtering
 */
public enum LSRouteArchetype: String, CaseIterable, Sendable {
    case all
    case twisties
    case scenic
    case technical
    case cruising
    case sport
    case adventure

    /// Display label for this archetype
    var label: String {
        switch self {
        case .all: return "All"
        case .twisties: return "Twisties"
        case .scenic: return "Scenic"
        case .technical: return "Technical"
        case .cruising: return "Cruising"
        case .sport: return "Sport"
        case .adventure: return "Adventure"
        }
    }

    /// Icon name for this archetype (optional)
    var iconName: String? {
        switch self {
        case .all: return "checkmark.circle.fill"
        case .twisties: return "road.variant"
        case .scenic: return "landscape.fill"
        case .technical: return "wrench.fill"
        case .cruising: return "bicycle"
        case .sport: return "flame.fill"
        case .adventure: return "compass.fill"
        }
    }
}

// MARK: - Discovery Filter Bar Component

/**
 * Discovery filter bar molecule component
 *
 * Horizontal scrollable archetype filter chips for route discovery.
 * Glassmorphic design with semi-transparent background.
 * Following React Native component from react-native/components/discovery/discovery-filter-bar.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Background: `theme.colors.surface.default` with 0.8 opacity
 *   - Border: `theme.colors.border.default` with 0.2 opacity
 * - Layout:
 *   - Padding: `theme.space.md` top/bottom, `theme.space.lg` horizontal
 *   - Chip spacing: `theme.space.sm`
 * - Glassmorphic effect: Semi-transparent surface with bottom border
 *
 * ## Behavior
 * - "All" chip clears the filter (sets selection to empty)
 * - Other archetypes toggle selection
 * - Deselecting last archetype shows all (clears filter)
 * - Count badges display formatted (e.g., "Scenic (12)", "Twisties (99+)", "All (1.2k)")
 *
 * ## Parameters
 * - selectedArchetypes: Array of currently selected archetypes
 * - onArchetypeChange: Callback when selection changes
 * - counts: Dictionary of archetype counts to display
 * - testID: Optional testing identifier for UI tests
 */
public struct LSDiscoveryFilterBar: View {
    @Environment(\.theme) private var theme

    private let selectedArchetypes: [LSRouteArchetype]
    private let onArchetypeChange: ([LSRouteArchetype]) -> Void
    private let counts: [LSRouteArchetype: Int]
    private let testID: String

    public init(
        selectedArchetypes: [LSRouteArchetype],
        onArchetypeChange: @escaping ([LSRouteArchetype]) -> Void,
        counts: [LSRouteArchetype: Int],
        testID: String = "discovery-filter-bar"
    ) {
        self.selectedArchetypes = selectedArchetypes
        self.onArchetypeChange = onArchetypeChange
        self.counts = counts
        self.testID = testID
    }

    // MARK: - Body

    public var body: some View {
        VStack(spacing: 0) {
            // Glassmorphic background container
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: theme.space.sm) {
                    // "All" chip
                    archetypeChip(for: .all)

                    // Archetype chips (exclude "all" from iteration)
                    ForEach(LSRouteArchetype.allCases.filter { $0 != .all }, id: \.self) { archetype in
                        archetypeChip(for: archetype)
                    }
                }
                .padding(.horizontal, theme.space.lg)
                .padding(.vertical, theme.space.md)
            }
            .accessibilityElement(children: .contain)
            .accessibilityLabel("Route archetype filters")

            // Bottom border
            Rectangle()
                .fill(theme.colors.border.default.opacity(0.2))
                .frame(height: theme.borderWidth.thin)
        }
        .background(theme.colors.surface.default.opacity(0.8))
    }

    // MARK: - Archetype Chip

    @ViewBuilder
    private func archetypeChip(for archetype: LSRouteArchetype) -> some View {
        let isSelected = isSelected(archetype)
        let count = counts[archetype] ?? 0
        let label = "\(archetype.label) (\(formatCount(count)))"

        LSChip(
            label,
            selected: isSelected,
            onPress: {
                handleArchetypeTap(archetype)
            },
            icon: {
                if let iconName = archetype.iconName {
                    AnyView(Image(systemName: iconName))
                } else {
                    AnyView(EmptyView())
                }
            },
            testID: "\(testID)-chip-\(archetype.rawValue)"
        )
    }

    // MARK: - Selection Logic

    private func isSelected(_ archetype: LSRouteArchetype) -> Bool {
        if archetype == .all {
            return selectedArchetypes.isEmpty
        }
        return selectedArchetypes.contains(archetype)
    }

    private func handleArchetypeTap(_ archetype: LSRouteArchetype) {
        if archetype == .all {
            // "All" clears the filter
            onArchetypeChange([])
            return
        }

        // Toggle the archetype
        let isSelected = selectedArchetypes.contains(archetype)
        if isSelected {
            // If deselecting and this was the only one, show all
            if selectedArchetypes.count == 1 {
                onArchetypeChange([])
            } else {
                onArchetypeChange(selectedArchetypes.filter { $0 != archetype })
            }
        } else {
            // Add to selection
            onArchetypeChange(selectedArchetypes + [archetype])
        }
    }

    // MARK: - Count Formatting

    /**
     * Format count for display (e.g., 12, 99+, 1.2k)
     */
    private func formatCount(_ count: Int) -> String {
        if count >= 1000 {
            return String(format: "%.1fk", Double(count) / 1000.0)
        }
        if count > 99 {
            return "99+"
        }
        return String(count)
    }
}

// MARK: - Test Helper

extension LSDiscoveryFilterBar {
    /**
     * Test helper to simulate tapping an archetype chip
     * Used in unit tests to verify selection behavior
     */
    func simulateTap(archetype: LSRouteArchetype) {
        handleArchetypeTap(archetype)
    }
}

// MARK: - Preview

#Preview("Discovery Filter Bar") {
    VStack(alignment: .leading, spacing: 24) {
        Text("Discovery Filter Bar")
            .font(.title)
            .fontWeight(.bold)
            .padding(.horizontal)

        VStack(alignment: .leading, spacing: 8) {
            Text("No Selection (All selected)")
                .font(.headline)
                .padding(.horizontal)

            LSDiscoveryFilterBar(
                selectedArchetypes: [],
                onArchetypeChange: { archetypes in
                    print("Selected: \(archetypes)")
                },
                counts: [
                    .all: 1250,
                    .scenic: 250,
                    .twisties: 300,
                    .technical: 150,
                    .cruising: 200,
                    .sport: 100,
                    .adventure: 50
                ]
            )
        }

        VStack(alignment: .leading, spacing: 8) {
            Text("Multiple Selection")
                .font(.headline)
                .padding(.horizontal)

            LSDiscoveryFilterBar(
                selectedArchetypes: [.scenic, .twisties],
                onArchetypeChange: { archetypes in
                    print("Selected: \(archetypes)")
                },
                counts: [
                    .all: 1250,
                    .scenic: 250,
                    .twisties: 300,
                    .technical: 150,
                    .cruising: 200,
                    .sport: 100,
                    .adventure: 50
                ]
            )
        }

        VStack(alignment: .leading, spacing: 8) {
            Text("Single Selection")
                .font(.headline)
                .padding(.horizontal)

            LSDiscoveryFilterBar(
                selectedArchetypes: [.technical],
                onArchetypeChange: { archetypes in
                    print("Selected: \(archetypes)")
                },
                counts: [
                    .all: 1250,
                    .scenic: 250,
                    .twisties: 300,
                    .technical: 150,
                    .cruising: 200,
                    .sport: 100,
                    .adventure: 50
                ]
            )
        }

        Spacer()
    }
    .laneShadowTheme()
}
